Empty Adoptium/adoptopenjdk Java list

### Service(s)

Update center

### Summary

This relates to an issue in the crawler as mentioning in Gitter

Since 13/03/26 the Jenkins tool installer has failed for any project using Adoptium/AdoptOpenJDK

The Adoptium installer JSON has been corrupted on at least what it seems to resolve to for me https://westeurope.cloudflare.jenkins.io/updates/io.jenkins.plugins.adoptopenjdk.AdoptOpenJDKInstaller.json. The raw API for Adoptium works fine for me locally from four different networks.

dduportal said they saw this
`java.lang.AssertionError: Size of target/io.jenkins.plugins.adoptopenjdk.AdoptOpenJDKInstaller.json was 110, less than minimum 300000`

I was testing out general connectivity to the API using `curl -sL "https://api.adoptium.net/v3/assets/feature_releases/${RELNUM}/ga?image_type=jdk" | jq -r '.[].binaries[] | select((.os == "windows" or .os == "linux") and (.architecture == "x64" or .architecture == "aarch64")) | .package.link | select(endswith(".zip") or endswith(".tar.gz"))' | head -n 3` but upon RE-trying today the first invocation hung. The second returned a 502. And then every subsequent request has been fine. 

I note the first time I change the RELNUM from, say, 21 to 25 the first request takes a second or so longer and subsequent are near instant. That suggests to me there is perhaps some kind of proxy caching layer and perhaps either a proxy in a bad state on the Adoptium side which perhaps is what the crawler first saw.

But if this was the case, in looking at the code referred to me there: https://github.com/jenkins-infra/crawler/blob/f9bda8a13339cf34d9369e8bbd9ade6bbdd76b5a/adoptopenjdk.groovy I see that the handling of foul weather cases doesn't seem to exist. A hanging (eventually timing out) request or a 502 would seemingly just return nothing, whilst any exception would be swallowed. 

Given the external resource dependency perhaps a few simple delayed rechecks on timedout/otherwise failing requests would address the general case.

I've mentioned this observation in the Adoptium Slack channel (which by chance I had open and saw someone else report this) but regardless of any fixing of the infra, resiliency in the face of external service flakiness would be a better option I think.

The current state of the service API is confusing to me. Adapting the query used in the crawler code I see varying results that don't make a lot of sense:

For example, this works with RELNUM of 11 and 17. But not 21 or 25. If I remove the page_size query param it works for 21 now, but still not 25. 
`curl -sL "[https://api.adoptium.net/v3/assets/feature_releases/${RELNUM}/ga?vendor=eclipse&project=jdk&image_type=jdk&page_size=20&page=1](https://api.adoptium.net/v3/assets/feature_releases/$%7BRELNUM%7D/ga?vendor=eclipse&project=jdk&image_type=jdk&page_size=20&page=1)"`



### Reproduction steps

https://westeurope.cloudflare.jenkins.io/updates/io.jenkins.plugins.adoptopenjdk.AdoptOpenJDKInstaller.json is empty, which is the reproduction from the client side.

In terms of the conditions to be able to handle, right now I can't run this with RELNUM of `21` or `25` but it works for `11` and `17`

`curl -sL "[https://api.adoptium.net/v3/assets/feature_releases/${RELNUM}/ga?vendor=eclipse&project=jdk&image_type=jdk&page_size=20&page=1](https://api.adoptium.net/v3/assets/feature_releases/$%7BRELNUM%7D/ga?vendor=eclipse&project=jdk&image_type=jdk&page_size=20&page=1)"`

> The Adoptium installer JSON has been corrupted on at least what it seems to resolve to for me https://westeurope.cloudflare.jenkins.io/updates/io.jenkins.plugins.adoptopenjdk.AdoptOpenJDKInstaller.json. The raw API for Adoptium works fine for me locally from four different networks.
> 
> dduportal said they saw this `java.lang.AssertionError: Size of target/io.jenkins.plugins.adoptopenjdk.AdoptOpenJDKInstaller.json was 110, less than minimum 300000`

I download a file of size 335883 with the command:

`wget https://westeurope.cloudflare.jenkins.io/updates/io.jenkins.plugins.adoptopenjdk.AdoptOpenJDKInstaller.json`

I suspect that means the Eclipse API was returning incorrect results and now it is returning correct results.

The general recommendation from the [Eclipse Temurin installer plugin documentation](https://plugins.jenkins.io/adoptopenjdk/#plugin-content-usage-recommendations) is:

> We want to warn that this plugin is NOT a good practice for production environments. As it relies on Adoptium's website to do the job, it's highly likely to stop working. It could happen because Adoptium's website changes or even if Adoptium bans our downloads due to excessive bandwidth usage or some other reason. Currently Adoptium is using GitHub for hosting release archives and GitHub could also stop working due to similar reasons.
>
> The recommended and preferred approach is to download the JDK distribution using other installers, for example downloading it from a well known URL (preferably hosted on your own network) with ZIP Tool Installer, having it pre-installed in agent docker images, or executing a script to do the job.

That recommendation does not excuse the error, only notes that there are techniques to make the error much less relevant in your installation.

I've configured my agents to already have the expected JDK installed and then I use the Eclipse Temurin installer plugin as the last of the installers in case an agent is not correctly configured with the expected JDK installed.  My [tool installer definition](https://github.com/MarkEWaite/docker-lfs/blob/lts-with-plugins/ref/configuration-as-code/tool.yaml) provides a reference implementation for configuration as code users ([commit fd01c65](https://github.com/MarkEWaite/docker-lfs/blob/fd01c65dadba5472ef0b042446ef46b8a87fc219/ref/configuration-as-code/tool.yaml#L170)).

--- a/.jenkins-scripts/generate.sh
+++ b/.jenkins-scripts/generate.sh
@@ -12,8 +12,7 @@ fi
 for f in [a-z]*.groovy
 do
     echo "= Crawler '$f':"
-    groovy -Dgrape.config=./grapeConfig.xml ./lib/runner.groovy "$f" \
-      || true # Hide failures and allow all generators to run
+    groovy -Dgrape.config=./grapeConfig.xml ./lib/runner.groovy "$f"
 done
 
 # Run a sanity test of the outputs

--- a/ZenithTest.groovy
+++ b/ZenithTest.groovy
@@ -185,6 +185,6 @@ class ZenithTest {
 
     @Test
     void terraform() {
-        checkFileSize(9_700_000L, "org.jenkinsci.plugins.terraform.TerraformInstaller.json");
+        checkFileSize(700_000L, "org.jenkinsci.plugins.terraform.TerraformInstaller.json");
     }
 }

--- a/ZenithTest.groovy
+++ b/ZenithTest.groovy
@@ -4,13 +4,24 @@ import static org.junit.Assert.assertTrue;
 
 import java.io.File;
 import org.junit.Ignore;
+import org.junit.Rule;
 import org.junit.Test;
+import org.junit.rules.TestRule;
+import org.junit.rules.TestWatcher;
 
 /**
  * Test to be run after all installer metadata has been generated.
  * Sanity checks for the installer metadata.
  */
 class ZenithTest {
+
+    @Rule
+    public TestRule watchman = [failed: {e, d ->
+                println d
+                e.printStackTrace()
+                System.exit(1)
+                }
+        ] as TestWatcher; // System.exit(1) if any test fails
 
     void checkFileSize(long minimumSize, String fileName) {
         File dataFile = new File("target/" + fileName);
@@ -19,7 +30,7 @@ class ZenithTest {
         long actualSize = dataFile.length();
         assertTrue(
                 "Size of target/" + fileName + " was " + actualSize + ", less than minimum " + minimumSize,
-                actualSize > minimumSize,
+                actualSize >= minimumSize,
                 );
     }
 

--- a/ZenithTest.groovy
+++ b/ZenithTest.groovy
@@ -174,6 +174,6 @@ class ZenithTest {
 
     @Test
     void terraform() {
-        checkFileSize(70_000L, "org.jenkinsci.plugins.terraform.TerraformInstaller.json");
+        checkFileSize(9_700_000L, "org.jenkinsci.plugins.terraform.TerraformInstaller.json");
     }
 }

